File: frontend/src/shared/types/course.ts
import type { LevelResponseDTO } from './level'
import type { UserResponseDTO } from './user'

export type CourseRequestDTO = {
	title: string
	description?: string
}

export type CourseResponseDTO = {
	id: number
	title: string
	description?: string
	author: UserResponseDTO
	levels: LevelResponseDTO[]
	participants: UserResponseDTO[]
	createdAt: string
}

export type CourseJoinDTO = {
	courseId: number
}

export type CourseSearchParams = {
	query: string
}

export type CourseUserProgressDTO = {
  user: UserResponseDTO;
  averageProgress: number;
};

File: dictionary/src/main/java/org/ru/dictionary/entity/Course.java
package org.ru.dictionary.entity;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.OnDelete;
import org.hibernate.annotations.OnDeleteAction;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

@Entity
@Table(indexes = @Index(columnList = "author_id"))
@Data
@NoArgsConstructor
@Builder
@AllArgsConstructor
public class Course {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "author_id", nullable = false)
    @ToString.Exclude
    @EqualsAndHashCode.Exclude
    private User author;

    @Column(nullable = false)
    private String title;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(name = "imagePath")
    private String imagePath;

    @ManyToMany(cascade = {CascadeType.PERSIST, CascadeType.MERGE})
    @JoinTable(
            name = "course_participants",
            joinColumns = @JoinColumn(name = "course_id"),
            inverseJoinColumns = @JoinColumn(name = "user_id")
    )
    private Set<User> participants = new HashSet<>();

    @OneToMany(
            mappedBy = "course",
            cascade = CascadeType.ALL, // Каскадное удаление
            orphanRemoval = true       // Удаление "осиротевших" уровней
    )
    @OrderBy("orderNumber ASC")
    @OnDelete(action = OnDeleteAction.CASCADE)
    private List<Level> levels = new ArrayList<>();

    @CreationTimestamp
    private LocalDateTime createdAt;
}